feat(examples): Ivan's registry, home realm

Ursulovic opened the pull request:
Created registry for further development of realms, finalising home realm.

Gno2D2 commented:
#### 🛠 PR Checks Summary
All **Automated Checks** passed. ✅

##### Manual Checks (for Reviewers):
- [ ] **SKIP**: Do not block the CI for this PR
- [ ] The pull request description provides enough details


<details><summary>Read More</summary>

🤖 This bot helps streamline PR reviews by verifying automated checks and providing guidance for contributors and reviewers.

##### ✅ Automated Checks (for Contributors):
 🟢 Maintainers must be able to edit this pull request ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))
 🟢 The pull request head branch must be up-to-date with its base ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/keeping-your-pull-request-in-sync-with-the-base-branch))


##### ☑️ Contributor Actions:
1. Fix any issues flagged by automated checks.
2. Follow the Contributor Checklist to ensure your PR is ready for review.
	- Add new tests, or document why they are unnecessary.
	- Provide clear examples/screenshots, if necessary.
	- Update documentation, if required.
	- Ensure no breaking changes, or include `BREAKING CHANGE` notes.
	- Link related issues/PRs, where applicable.

##### ☑️ Reviewer Actions:
1. Complete manual checks for the PR, including the guidelines and additional checks if applicable.

##### 📚 Resources:
- [Report a bug with the bot](https://github.com/gnolang/gno/issues/3238).
- [View the bot’s configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 The pull request was created from a fork (head branch repo: Ursulovic/gno)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Maintainer can modify this pull request

```
</blockquote></details>

<details><summary>The pull request head branch must be up-to-date with its base (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Head branch (Ursulovic:home-page) is up to date with base (master): behind by 0 / ahead by 1

```
</blockquote></details>

</blockquote></details>


<details><summary><b>Manual Checks</b></summary><blockquote>

<details><summary>**SKIP**: Do not block the CI for this PR</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Can be checked by

- Any user with comment edit permission

</blockquote></details>

<details><summary>The pull request description provides enough details</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 Not (🔴 Pull request author is a member of the team: core-contributors)

```
### Can be checked by
 - team core-contributors

</blockquote></details>

</blockquote></details>

</blockquote></details>

</details>

leohhhn commented:
When creating a PR, create it as Draft first 🙏

leohhhn commented:
Also you say finalizing, but your home realm package is empty 👀

moul commented on examples/gno.land/r/ursulovic/registry/registry.gno:
```diff
@@ -0,0 +1,71 @@
+		return ErrInvalidAddr
+	}
+
+	if err := checkAuthorized; err != nil {
```
```suggestion
	if err := checkAuthorized(); err != nil {
```

Ursulovic replied:
Fixed!

moul commented on examples/gno.land/r/ursulovic/registry/registry.gno:
```diff
@@ -0,0 +1,71 @@
+	return nil
+}
+
+func assertAuthorized() {
```
unused.

Ursulovic replied:
Fixed!

moul commented on examples/gno.land/r/ursulovic/registry/registry.gno:
```diff
@@ -0,0 +1,71 @@
+	mainAddress   std.Address
+	backupAddress std.Address
+
+	ErrInvalidAddr  = errors.New("leon's config: invalid address")
```
```suggestion
	ErrInvalidAddr  = errors.New("ursulovic's config: invalid address")
```

Ursulovic replied:
Fixed!

Ursulovic commented:
Thank you @moul, I have implemented all your suggestions!



> Also you say finalizing, but your home realm package is empty 👀

Sorry, I forgot to stage the files earlier :/. I’ve added them now. I have created the Realm package, which primarily displays random blockchain "wisdom" and provides some information about me. Tomorrow, I plan to implement functionality  of modifying the realm data if  authorization is passed using my registry realm.

leohhhn commented on examples/gno.land/r/ursulovic/home/home.gno:
```diff
@@ -0,0 +1,90 @@
+	hof.Register()
+}
+
+//Update
```
If you want to write comments like this, I suggest you follow [godoc](https://go.dev/blog/godoc).

If you want to separate functions that are about "updating" things, write a comment 2 lines above the section. 
You can try using `goland` as an IDE :)

leohhhn reviewed (changes requested):
Thank you for this PR.

While this is good work, I suggest you try to be a bit more creative with your home realm. 

PRs in the `monorepo/examples` should be a good example and unique to newcomers. While this is good code, if you take a look the home realms of other people, this matches some things we have already seen. 

Please, try to find some interesting example realm, or make something completely new which can be imported into your realm to make it more unique. 

If you do not do this and want to keep your home realm like this, you can simply deploy it permissionlessly with `gnokey` or Adena.

Ursulovic commented:
Thank you for the update @leohhhn, I will think of new Home Realm idea and write new version soon!

leohhhn commented:
Seems that some code is still not properly formatted; as pointed out by the CI; you can run `make fmt` defined in the `examples/Makefile`. This will recusrively run `gno fmt` on all code in the folder.

Ursulovic commented:
> Seems that some code is still not properly formatted; as pointed out by the CI; you can run `make fmt` defined in the `examples/Makefile`. This will recusrively run `gno fmt` on all code in the folder.

Fixed formatting issues on this commit : https://github.com/gnolang/gno/pull/3354/commits/706784edc02d51ad344d0b1da97fdf7b4d3ee75e

leohhhn commented on examples/gno.land/r/ursulovic/home/home.gno:
```diff
@@ -0,0 +1,160 @@
+	return sb.String()
+}
+
+func initIsValidURL() {
+	isValidURL = func(url string) bool {
+		const urlPrefix string = "https://i.ibb.co/"
+
+		if !strings.HasPrefix(url, urlPrefix) {
+			return false
+		}
+
+		if !(strings.HasSuffix(url, ".jpg") ||
+			strings.HasSuffix(url, ".png") ||
+			strings.HasSuffix(url, ".gif") ||
+			strings.HasSuffix(url, ".webp")) {
+			return false
+		}
+
+		urlPath := strings.TrimPrefix(url, "https://i.ibb.co/")
+		parts := strings.Split(urlPath, "/")
+
+		if len(parts) != 2 || len(parts[0]) == 0 || len(parts[1]) == 0 {
+			return false
+		}
+
+		return true
+	}
+}
```
Nice try; you can just create a `func defaultURLValidation() {...}` and set `isValidURL = defaultURLValidation` at declaration

Ursulovic replied:
I implemented your suggestion in this commit: https://github.com/gnolang/gno/pull/3354/commits/1065ad29149f5dd48dcfb90131dadc9500c11175

leohhhn commented on examples/gno.land/r/ursulovic/home/home.gno:
```diff
@@ -0,0 +1,158 @@
+var (
+	aboutMe       string
+	selectedImage string
+	owner         *ownable.Ownable
+
+	githubUrl        string
+	linkedinUrl      string
+	imageUpdatePrice int64
+
+	isValidURL func(string) bool
+)
+
+func init() {
+	owner = ownable.NewWithAddress(registry.MainAddress())
```
Just one final thing I noticed; you should export Owner. This way the methods on the Ownable struct are exposed and callable from the outside, which is not the case currently. With your code right now, there is no way to `TransferOwnership` or `DropOwnership`. 

```suggestion
	Ownable         *ownable.Ownable

	githubUrl        string
	linkedinUrl      string
	imageUpdatePrice int64

	isValidURL func(string) bool
)

func init() {
	Ownable = ownable.NewWithAddress(registry.MainAddress())
```

Ursulovic replied:
I implemented these changes in commit : https://github.com/gnolang/gno/pull/3354/commits/0bfa81e66980d1fc0aaaf0292cab7fd7f40bcf0f.

Meanwhile I have also made another commit before this one in which I have made a link to connect studio updatable: https://github.com/gnolang/gno/pull/3354/commits/0bfa81e66980d1fc0aaaf0292cab7fd7f40bcf0f

leohhhn commented on examples/gno.land/r/ursulovic/home/home.gno:
```diff
@@ -33,7 +34,8 @@ func init() {
 	imageUpdatePrice = 5000000
-	isValidURL = defaultURLValidation
+	isValidUrl = defaultURLValidation
+	studioConnectUrl = "http://www.google.com"
```
Why is it "studio"?

Ursulovic replied:
> Why is it "studio"?

My bad, I though Connect was called Studio Connect, I renamed variable and function for updating it in this commit: https://github.com/gnolang/gno/pull/3354/commits/bb7676a1beaaf50083aa35fc5eeee7678b842276

leohhhn replied:
Sorry, I still don't fully undestand. You want to make a Connect URL to the `UpdateImage` function, right?

First, just leave the initial url at initialization empty; second, even though the link does not exist yet, you can guess it, no?
